File: backend/app/interfaces/schemas/node.py
from datetime import datetime
from typing import List, Optional, Literal

from pydantic import BaseModel, Field

from app.domain.models.node import SSHNode, SSHOperationLog, SSHCommandApproval, SSHApprovalStatus


class ServerNodeBase(BaseModel):
    name: str = Field(min_length=1, max_length=64)
    description: Optional[str] = Field(default=None, max_length=500)
    remarks: Optional[str] = Field(default=None, max_length=1000)

    ssh_enabled: bool = False
    ssh_host: Optional[str] = Field(default=None, max_length=255)
    ssh_port: int = Field(default=22, ge=1, le=65535)
    ssh_username: Optional[str] = Field(default=None, max_length=128)
    ssh_auth_type: Literal["password", "private_key"] = "password"
    ssh_password: Optional[str] = Field(default=None, max_length=2048)
    ssh_private_key: Optional[str] = Field(default=None, max_length=20000)
    ssh_passphrase: Optional[str] = Field(default=None, max_length=2048)
    ssh_require_approval: bool = False


class CreateServerNodeRequest(ServerNodeBase):
    pass


class UpdateServerNodeRequest(ServerNodeBase):
    pass


class ServerNodeResponse(BaseModel):
    node_id: str
    name: str
    description: Optional[str] = None
    remarks: Optional[str] = None

    ssh_enabled: bool
    ssh_host: Optional[str] = None
    ssh_port: int
    ssh_username: Optional[str] = None
    ssh_auth_type: Literal["password", "private_key"]
    ssh_password: Optional[str] = None
    ssh_private_key: Optional[str] = None
    ssh_passphrase: Optional[str] = None
    ssh_require_approval: bool

    created_at: datetime
    updated_at: datetime

    @classmethod
    def from_model(cls, node: SSHNode) -> "ServerNodeResponse":
        return cls(
            node_id=node.id,
            name=node.name,
            description=node.description,
            remarks=node.remarks,
            ssh_enabled=node.ssh_enabled,
            ssh_host=node.ssh_host,
            ssh_port=node.ssh_port,
            ssh_username=node.ssh_username,
            ssh_auth_type=node.ssh_auth_type.value,
            ssh_password=node.ssh_password,
            ssh_private_key=node.ssh_private_key,
            ssh_passphrase=node.ssh_passphrase,
            ssh_require_approval=node.ssh_require_approval,
            created_at=node.created_at,
            updated_at=node.updated_at,
        )


class ListServerNodesResponse(BaseModel):
    nodes: List[ServerNodeResponse]


class SSHExecRequest(BaseModel):
    command: str = Field(min_length=1, max_length=8000)
    exec_dir: Optional[str] = None
    sync_to_ai: bool = False
    session_id: Optional[str] = None


class SSHExecResponse(BaseModel):
    success: bool
    output: str
    command: str
    node_id: str
    node_name: str


class SSHMonitorResponse(BaseModel):
    node_id: str
    node_name: str
    system_info: str


class SSHLogItem(BaseModel):
    log_id: str
    session_id: Optional[str]
    actor_type: str
    actor_id: Optional[str] = None
    source: str
    command: str
    output: str
    success: bool
    created_at: datetime

    @classmethod
    def from_model(cls, log: SSHOperationLog) -> "SSHLogItem":
        return cls(
            log_id=log.id,
            session_id=log.session_id,
            actor_type=log.actor_type,
            actor_id=log.actor_id,
            source=log.source,
            command=log.command,
            output=log.output,
            success=log.success,
            created_at=log.created_at,
        )


class SSHLogsResponse(BaseModel):
    logs: List[SSHLogItem]


class PendingApprovalItem(BaseModel):
    approval_id: str
    session_id: str
    node_id: str
    command: str
    status: SSHApprovalStatus
    reject_reason: Optional[str]
    created_at: datetime

    @classmethod
    def from_model(cls, model: SSHCommandApproval) -> "PendingApprovalItem":
        return cls(
            approval_id=model.id,
            session_id=model.session_id,
            node_id=model.node_id,
            command=model.command,
            status=model.status,
            reject_reason=model.reject_reason,
            created_at=model.created_at,
        )


class PendingApprovalListResponse(BaseModel):
    approvals: List[PendingApprovalItem]


class DecideApprovalRequest(BaseModel):
    approve: bool
    reject_reason: Optional[str] = None


class DecideApprovalResponse(BaseModel):
    approval_id: str
    status: SSHApprovalStatus
    output: Optional[str] = None

File: frontend/src/api/node.ts
import { apiClient, ApiResponse } from './client';

export interface ServerNode {
  node_id: string;
  name: string;
  description?: string;
  remarks?: string;
  ssh_enabled: boolean;
  ssh_host?: string;
  ssh_port: number;
  ssh_username?: string;
  ssh_auth_type: 'password' | 'private_key';
  ssh_password?: string;
  ssh_private_key?: string;
  ssh_passphrase?: string;
  ssh_require_approval: boolean;
  created_at: string;
  updated_at: string;
}

export interface ListServerNodesResponse {
  nodes: ServerNode[];
}

export interface SSHExecResponse {
  success: boolean;
  output: string;
  command: string;
  node_id: string;
  node_name: string;
}

export interface SSHMonitorResponse {
  node_id: string;
  node_name: string;
  system_info: string;
}

export interface NodeOverviewMetric {
  label: string;
  value: string;
  hint?: string;
  level: 'ok' | 'warn' | 'critical';
}

export interface NodeOverviewResponse {
  node_id: string;
  node_name: string;
  checked_at: string;
  status: 'healthy' | 'warning' | 'critical';
  summary: string;
  hostname?: string;
  os_name?: string;
  kernel?: string;
  uptime?: string;
  load_average?: string;
  memory_total?: string;
  memory_used?: string;
  memory_free?: string;
  disk_total?: string;
  disk_used?: string;
  disk_use_percent?: number;
  metrics: NodeOverviewMetric[];
  raw_output: string;
}

export interface SSHLogItem {
  log_id: string;
  session_id?: string;
  actor_type: string;
  actor_id?: string;
  source: string;
  command: string;
  output: string;
  success: boolean;
  created_at: string;
}

export interface SSHLogsResponse {
  logs: SSHLogItem[];
}

export interface PendingApprovalItem {
  approval_id: string;
  session_id: string;
  node_id: string;
  command: string;
  status: 'pending' | 'approved' | 'rejected';
  reject_reason?: string;
  created_at: string;
}

export interface PendingApprovalListResponse {
  approvals: PendingApprovalItem[];
}

export interface DecideApprovalResponse {
  approval_id: string;
  status: 'pending' | 'approved' | 'rejected';
  output?: string;
}

export async function listServerNodes(): Promise<ServerNode[]> {
  const response = await apiClient.get<ApiResponse<ListServerNodesResponse>>('/nodes');
  return response.data.data.nodes;
}

export async function createServerNode(payload: Omit<ServerNode, 'node_id' | 'created_at' | 'updated_at'>): Promise<ServerNode> {
  const response = await apiClient.post<ApiResponse<ServerNode>>('/nodes', payload);
  return response.data.data;
}

export async function updateServerNode(nodeId: string, payload: Omit<ServerNode, 'node_id' | 'created_at' | 'updated_at'>): Promise<ServerNode> {
  const response = await apiClient.put<ApiResponse<ServerNode>>(`/nodes/${nodeId}`, payload);
  return response.data.data;
}

export async function deleteServerNode(nodeId: string): Promise<void> {
  await apiClient.delete<ApiResponse<void>>(`/nodes/${nodeId}`);
}

export async function execNodeSSH(nodeId: string, command: string, execDir?: string, syncToAi?: boolean, sessionId?: string): Promise<SSHExecResponse> {
  const response = await apiClient.post<ApiResponse<SSHExecResponse>>(`/nodes/${nodeId}/ssh/exec`, {
    command,
    exec_dir: execDir,
    sync_to_ai: !!syncToAi,
    session_id: sessionId,
  });
  return response.data.data;
}

export async function monitorNode(nodeId: string): Promise<SSHMonitorResponse> {
  const response = await apiClient.get<ApiResponse<SSHMonitorResponse>>(`/nodes/${nodeId}/monitor`);
  return response.data.data;
}

export async function getNodeOverview(nodeId: string): Promise<NodeOverviewResponse> {
  const response = await apiClient.get<ApiResponse<NodeOverviewResponse>>(`/nodes/${nodeId}/overview`);
  return response.data.data;
}

export async function listNodeLogs(nodeId: string, limit: number = 100, includeSystem: boolean = false): Promise<SSHLogItem[]> {
  const response = await apiClient.get<ApiResponse<SSHLogsResponse>>(`/nodes/${nodeId}/logs`, {
    params: { limit, include_system: includeSystem },
  });
  return response.data.data.logs;
}

export async function listSessionApprovals(sessionId: string): Promise<PendingApprovalItem[]> {
  const response = await apiClient.get<ApiResponse<PendingApprovalListResponse>>(`/nodes/sessions/${sessionId}/approvals`);
  return response.data.data.approvals;
}

export async function decideApproval(approvalId: string, approve: boolean, rejectReason?: string): Promise<DecideApprovalResponse> {
  const response = await apiClient.post<ApiResponse<DecideApprovalResponse>>(`/nodes/approvals/${approvalId}/decision`, {
    approve,
    reject_reason: rejectReason,
  });
  return response.data.data;
}
